File: code/dynamic_programming/knapsack/Knapsack.java
// Part of Cosmos by OpenGenus Foundation

class Knapsack {
    public static void main(String[] args) throws Exception {
        int val[] = {10, 40, 30, 50};
        int wt[] = {5, 4, 6, 3};
        int W = 10;
        System.out.println(knapsack(val, wt, W));
    }
    public static int knapsack(int val[], int wt[], int W) {
        int N = wt.length; // Get the total number of items. Could be wt.length or val.length. Doesn't matter
        int[][] V = new int[N + 1][W + 1]; //Create a matrix. Items are in rows and weight at in columns +1 on each side
        //What if the knapsack's capacity is 0 - Set all columns at row 0 to be 0
        for (int col = 0; col <= W; col++) {
            V[0][col] = 0;
        }
        //What if there are no items at home.  Fill the first row with 0
        for (int row = 0; row <= N; row++) {
            V[row][0] = 0;
        }
        for (int item=1;item<=N;item++){
            //Let's fill the values row by row
            for (int weight=1;weight<=W;weight++){
                //Is the current items weight less than or equal to running weight
                if (wt[item-1]<=weight){
                    //Given a weight, check if the value of the current item + value of the item that we could afford with the remaining weight
                    //is greater than the value without the current item itself
                    V[item][weight]=Math.max (val[item-1]+V[item-1][weight-wt[item-1]], V[item-1][weight]);
                }
                else {
                    //If the current item's weight is more than the running weight, just carry forward the value without the current item
                    V[item][weight]=V[item-1][weight];
                }
            }
        }
        //Printing the matrix
        for (int[] rows : V) {
            for (int col : rows) {
                System.out.format("%5d", col);
            }
            System.out.println();
        }
        return V[N][W];
    }
}

File: code/dynamic_programming/src/knapsack/knapsack.js
/**
 * Part of Cosmos by OpenGenus Foundation
 */

/**
 * method which returns the maximum total value in the knapsack
 * @param {array} valueArray
 * @param {array} weightArray
 * @param {integer} maximumWeight
 */
function knapsack(valueArray, weightArray, maximumWeight) {
  let n = weightArray.length;
  let matrix = [];

  // lazy initialize the element of 2d array
  if (!matrix[0]) matrix[0] = [];

  // if the knapsack's capacity is 0 - Set all columns at row 0 to be 0
  for (let inc = 0; inc <= maximumWeight; inc++) {
    matrix[0][inc] = 0;
  }

  for (let i = 1; i < n + 1; i++) {
    // check all possible maxWeight values from 1 to W
    for (let j = 1; j < maximumWeight + 1; j++) {
      // lazy initialize the element of 2d array
      if (!matrix[j]) matrix[j] = [];

      // if current item weighs < j then we have two options,
      if (weightArray[i - 1] <= j) {
        // Given a weight, check if the value of the current item + value of the item that we could afford with the remaining weight
        // is greater than the value without the current item itself
        matrix[i][j] = Math.max(
          matrix[i - 1][j],
          valueArray[i - 1] + matrix[i - 1][j - weightArray[i - 1]]
        );
      } else {
        // If the current item's weight is more than the running weight, just carry forward the value without the current item
        matrix[i][j] = matrix[i - 1][j];
      }
    }
  }

  return matrix[n][maximumWeight];
}

console.log(
  "Result - " + knapsack([12, 1000, 30, 10, 1000], [19, 120, 20, 1, 120], 40)
);
